### ISSUE TITLE
How to get the images which gave FP and FN from evaluation/test dataset?

### ISSUE THREAD

**Issue by noobing4Fun** [ultralytics/ultralytics#3342]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have successfully trained and validated my first YOLOv8 object detection model.

I now want to view which of the test images contained detections that had false positives and false negatives so that I can investigate these further. I guess I could manually calculate these using the _predictions_ file, but wanted to know, is there an existing Ultralytics tool or function that already exists for this?  I

**Comment by glenn-jocher** [comment]
@noobing4Fun hi,

Thank you for using YOLOv8! Currently, there is no specific tool or function in YOLOv8 to directly retrieve the images that had false positives (FP) or false negatives (FN) during evaluation or testing.

However, you can manually calculate these using the predictions file. The predictions file contains the bounding box coordinates, confidence scores, and class labels for each detected object in the test images. By comparing the predicted results with the ground truth annotations of the test images, you can identify the false positives and false negatives.

You can write a script or use a data analysis tool (e.g., Python with Pandas) to iterate through the predictions file and compare the predicted bounding boxes with the ground truth annotations. Based on the evaluation metrics and criteria you have set, you can identify the images that have false positives and false negatives.

Feel free to ask if you have any further questions or if there's anything else I can assist you with!

Best regards,
Glenn Jocher

**Comment by noobing4Fun** [comment]
Thank you for the feedback.

**Comment by glenn-jocher** [comment]
@noobing4Fun hi [username],

Thank you for your feedback! We appreciate your input and are constantly working to improve YOLOv8. Our team always strives to make the YOLOv8 repository a valuable resource for the community.

Regarding your suggestion of adding an option to visualize the ground truth bounding boxes during training, I can understand why you find this feature useful. At the moment, YOLOv8 does not have a built-in option specifically for visualizing ground truth bounding boxes during training.

However, you can achieve this by modifying the code slightly. After loading the batch of images and labels in the training loop, you can use a plotting library like OpenCV or Matplotlib to draw the ground truth bounding boxes on the images before passing them to the model. This way, you can visualize the ground truth boxes alongside the predicted boxes during training.

We greatly value user feedback, and your suggestion will certainly be taken into consideration for future updates to YOLOv8. If you have any further questions or suggestions, please don't hesitate to let us know.

Thank you again for your contribution!

Best regards,
Glenn Jocher

**Comment by dimpolitik** [comment]
A nice tool to explore predictions after training is FiftyOne.

https://docs.voxel51.com/

**Comment by glenn-jocher** [comment]
@dimpolitik hello! To review the images with false positives (FP) and false negatives (FN) from your evaluation/test dataset, you can use the `--save-hybrid` flag during validation. This will save a hybrid version of your labels that includes both ground truth and model predictions. Here's how you can do it:

```bash
yolo detect val model=your_model.pt data=your_data.yaml save_hybrid=True
```

After running this command, you'll find the saved images in the `runs/val/exp*/hybrid` directory. The images will have both ground truth and predicted bounding boxes, allowing you to visually inspect FPs and FNs.

For a more detailed analysis, you might need to write a custom script to parse the saved predictions and compare them with the ground truth annotations. Currently, there isn't a built-in Ultralytics tool that provides a list of images specifically filtered by FP and FN detections.

If you have further questions or need assistance with the script, feel free to reach out to the community in the [discussions](https://github.com/ultralytics/ultralytics/discussions) section. 😊🛠️

**Comment by dimpolitik** [comment]
@glenn-jocher Great update in the toolbox of Ultralytics. Not aware of that before. Thank you!!

**Comment by glenn-jocher** [comment]
@dimpolitik hello! Great to hear that you've successfully trained and validated your YOLOv8 model. 👍

Currently, YOLOv8 does not provide a built-in tool specifically for extracting images with false positives (FP) and false negatives (FN) directly from the validation process. However, you can achieve this by analyzing the saved prediction results.

During validation, if you set `save_json=True` and `save_hybrid=True` in your validation settings, YOLOv8 will save the predictions in JSON format and a hybrid version of labels (ground truth + predictions). You can then use these files to identify which images contain FPs and FNs by comparing the predictions with the ground truth annotations.

Here's a concise example of how you might enable these settings in Python:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/model.pt')

# Validate the model and save results
metrics = model.val(save_json=True, save_hybrid=True)
```

After running validation with these settings, you can write a custom script to parse the JSON file and identify the images with discrepancies between the ground truth and the predictions.

I hope this helps! If you have any more questions or need further assistance, feel free to ask. Happy investigating! 😊

**Comment by jamesquags** [comment]
this should really be a function of Ultralytics Explorer IMHO or somewhere within the Ultralytics World. 

integrated feedback on the images that are most challenging for the model or have breakdowns in ground truth will allow users to implement corrective action or immediately understand what type of images/labels they need more of.

I feel like it's super low hanging fruit to raise the average mAP of all custom models trained with yolov8 with this functionality!  if it's easier for users to find errors and correct them.  accuracy will go up! 😊

**Comment by jamesquags** [comment]
> @glenn-jocher Great update in the toolbox of Ultralytics. Not aware of that before. Thank you!!

what do you mean toolbox?

**Comment by glenn-jocher** [comment]
@jamesquags currently, there isn't a built-in Ultralytics tool that directly outputs the images with false positives (FP) and false negatives (FN) after validation. However, you can achieve this by analyzing the results of the validation process, which typically includes a JSON file with detailed information about the detections.

Here's a general approach to identify the images with FP and FN:

1. **Run Validation with JSON Output**: Ensure that you run the validation with the `--save-json` flag to get detailed output in JSON format. This will save a `.json` file with the predictions.

    ```bash
    yolo detect val model=yolov8n.pt data=coco128.yaml save_json=True
    ```

2. **Analyze the JSON File**: The JSON file contains detailed information about the predictions, including the image ID, predicted class, confidence score, and bounding box coordinates. You can parse this file to identify which predictions are FPs and FNs by comparing them with the ground truth annotations.

3. **Script for Analysis**: Write a Python script to parse the JSON file and the ground truth annotations. For each image, compare the predicted boxes with the ground truth. If a prediction doesn't match any ground truth box (based on IoU threshold), it's an FP. If a ground truth box doesn't have a matching prediction, it's an FN.

4. **Output Images with FP and FN**: Once you have identified the FPs and FNs, you can use the image IDs to retrieve the corresponding images from your dataset. You can then visualize these images with the predicted bounding boxes and ground truth to investigate further.

Here's a simplified example of how you might write a script to parse the JSON file and identify FPs and FNs:

```python
import json
from pathlib import Path

# Load predictions and ground truth
predictions_path = Path('runs/val/exp/predictions.json')
with open(predictions_path) as f:
    predictions = json.load(f)

# Assuming you have a function to load your ground truth annotations
# ground_truth = load_ground_truth_annotations()

# Placeholder for your ground truth loading function
def load_ground_truth_annotations():
    # Load your ground truth annotations here
    # This should return a dictionary with image IDs as keys and ground truth boxes as values
    pass

ground_truth = load_ground_truth_annotations()

# Analyze predictions
fp_images = set()
fn_images = set()

for pred in predictions:
    image_id = pred['image_id']
    gt_boxes = ground_truth.get(image_id, [])
    pred_box = [pred['bbox'], pred['score'], pred['category_id']]

    # Check for false positives and false negatives
    # This is a simplified check, you'll need to implement IoU matching and thresholding
    if not any(iou(pred_box, gt_box) > 0.5 for gt_box in gt_boxes):
        fp_images.add(image_id)
    if not any(iou(gt_box, pred_box) > 0.5 for pred_box in predictions if pred_box['image_id'] == image_id):
        fn_images.add(image_id)

# Output results
print(f"Images with false positives: {fp_images}")
print(f"Images with false negatives: {fn_images}")

# Implement the IoU function based on your requirements
def iou(boxA, boxB):
    # Calculate the intersection over union of two boxes
    pass
```

Remember to replace the placeholder functions with your actual data loading and IoU calculation logic. Once you have the sets of image IDs with FPs and FNs, you can visualize these images with the corresponding annotations to understand the model's mistakes better.

**Comment by ztahirzju** [comment]
> @jamesquags currently, there isn't a built-in Ultralytics tool that directly outputs the images with false positives (FP) and false negatives (FN) after validation. However, you can achieve this by analyzing the results of the validation process, which typically includes a JSON file with detailed information about the detections.
> 
> Here's a general approach to identify the images with FP and FN:
> 
> 1. **Run Validation with JSON Output**: Ensure that you run the validation with the `--save-json` flag to get detailed output in JSON format. This will save a `.json` file with the predictions.
>    ```shell
>    yolo detect val model=yolov8n.pt data=coco128.yaml save_json=True
>    ```
> 2. **Analyze the JSON File**: The JSON file contains detailed information about the predictions, including the image ID, predicted class, confidence score, and bounding box coordinates. You can parse this file to identify which predictions are FPs and FNs by comparing them with the ground truth annotations.
> 3. **Script for Analysis**: Write a Python script to parse the JSON file and the ground truth annotations. For each image, compare the predicted boxes with the ground truth. If a prediction doesn't match any ground truth box (based on IoU threshold), it's an FP. If a ground truth box doesn't have a matching prediction, it's an FN.
> 4. **Output Images with FP and FN**: Once you have identified the FPs and FNs, you can use the image IDs to retrieve the corresponding images from your dataset. You can then visualize these images with the predicted bounding boxes and ground truth to investigate further.
> 
> Here's a simplified example of how you might write a script to parse the JSON file and identify FPs and FNs:
> 
> ```python
> import json
> from pathlib import Path
> 
> # Load predictions and ground truth
> predictions_path = Path('runs/val/exp/predictions.json')
> with open(predictions_path) as f:
>     predictions = json.load(f)
> 
> # Assuming you have a function to load your ground truth annotations
> # ground_truth = load_ground_truth_annotations()
> 
> # Placeholder for your ground truth loading function
> def load_ground_truth_annotations():
>     # Load your ground truth annotations here
>     # This should return a dictionary with image IDs as keys and ground truth boxes as values
>     pass
> 
> ground_truth = load_ground_truth_annotations()
> 
> # Analyze predictions
> fp_images = set()
> fn_images = set()
> 
> for pred in predictions:
>     image_id = pred['image_id']
>     gt_boxes = ground_truth.get(image_id, [])
>     pred_box = [pred['bbox'], pred['score'], pred['category_id']]
> 
>     # Check for false positives and false negatives
>     # This is a simplified check, you'll need to implement IoU matching and thresholding
>     if not any(iou(pred_box, gt_box) > 0.5 for gt_box in gt_boxes):
>         fp_images.add(image_id)
>     if not any(iou(gt_box, pred_box) > 0.5 for pred_box in predictions if pred_box['image_id'] == image_id):
>         fn_images.add(image_id)
> 
> # Output results
> print(f"Images with false positives: {fp_images}")
> print(f"Images with false negatives: {fn_images}")
> 
> # Implement the IoU function based on your requirements
> def iou(boxA, boxB):
>     # Calculate the intersection over union of two boxes
>     pass
> ```
> 
> Remember to replace the placeholder functions with your actual data loading and IoU calculation logic. Once you have the sets of image IDs with FPs and FNs, you can visualize these images with the corresponding annotations to understand the model's mistakes better.

When I applied the same code: it showed me this error:

import json
from pathlib import Path

# Load predictions and ground truth
predictions_path = Path('/kaggle/working/runs/detect/val/predictions.json')
with open(predictions_path) as f:
    predictions = json.load(f)

# Assuming you have a function to load your ground truth annotations
# ground_truth = load_ground_truth_annotations()

# Placeholder for your ground truth loading function
def load_ground_truth_annotations():
    # Load your ground truth annotations here
    # This should return a dictionary with image IDs as keys and ground truth boxes as values
    pass

ground_truth = load_ground_truth_annotations()

# Analyze predictions
fp_images = set()
fn_images = set()

for pred in predictions:
    image_id = pred['image_id']
    gt_boxes = ground_truth.get(image_id, [])
    pred_box = [pred['bbox'], pred['score'], pred['category_id']]

    # Check for false positives and false negatives
    # This is a simplified check, you'll need to implement IoU matching and thresholding
    if not any(iou(pred_box, gt_box) > 0.5 for gt_box in gt_boxes):
        fp_images.add(image_id)
    if not any(iou(gt_box, pred_box) > 0.5 for pred_box in predictions if pred_box['image_id'] == image_id):
        fn_images.add(image_id)

# Output results
print(f"Images with false positives: {fp_images}")
print(f"Images with false negatives: {fn_images}")

# Implement the IoU function based on your requirements
def iou(boxA, boxB):
    # Calculate the intersection over union of two boxes
    pass

![Screenshot 2024-04-28 180225](https://github.com/ultralytics/ultralytics/assets/88961861/b95f47d7-0c75-44a8-980e-56d90e83937e)

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #18868, which introduces enhanced validation visualizations for Detect, Segment, Pose, and OBB tasks. With the new visualize option, users can now see ground truths, true positives, false positives, and false negatives per image—making debugging and understanding model performance much more intuitive. The update also improves internal handling of confusion matrices, updates related tests, and refreshes the documentation to help users get started with the new feature.

Special thanks to @Y-T-G for authoring the PR, and to @Laughing-q, @Skillnoob, @glenn-jocher, @Burhan-Q, @Petros626, and @O-J1 for their valuable contributions!

To test if these changes resolve the issue, try updating to the latest version with:
```
pip install -U ultralytics>=8.3.169
```
Alternatively, you can install directly from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Or, if you prefer, you can wait for the next official PyPI release.

Please let us know if this update addresses your issue! Thank you for reporting this, and feel free to share additional feedback if the problem persists. 🙏
